File: algorithm/python/BOJ17471.py
# from https://www.acmicpc.net/problem/17471

import sys

N = int(sys.stdin.readline())
people = list(map(int, sys.stdin.readline().split()))
people.insert(0, 0)
is_selected = [False] * (N + 1)
adj_matrix = []
result = sys.maxsize

for i in range(0, N + 1):
    adj_matrix.append([0] * (N + 1))

for i in range(1, N + 1):
    arr = list(map(int, sys.stdin.readline().split()))
    for j in range(1, arr[0] + 1):
        adj_matrix[i][arr[j]] = 1


def DFS(start, visited):
    # 선거구 지역별 연결 여부 확인
    visited[start] = True

    for i in range(1, N + 1):
        if not visited[i] and is_selected[start] == is_selected[i] and adj_matrix[start][i] == 1:
            DFS(i, visited)


def divide(cnt):
    # 선거구 나누기
    if cnt == N:
        flag = True
        for i in range(1, N):
            if is_selected[i] != is_selected[i + 1]:
                flag = False

        # 선거구를 한쪽으로만 나눴을 때 가지치기
        if flag:
            return

        for i in range(1, N + 1):
            if is_selected[i]:
                start_red = i
                break

        for i in range(1, N + 1):
            if not is_selected[i]:
                start_blue = i
                break

        visited = [False] * (N + 1)

        DFS(start_red, visited)
        DFS(start_blue, visited)

        for i in range(1, N + 1):
            if not visited[i]:
                return

        red_total = blue_total = 0

        for i in range(1, N + 1):
            if is_selected[i]:
                red_total += people[i]
            else:
                blue_total += people[i]

        global result
        if result > abs(red_total - blue_total):
            result = abs(red_total - blue_total)

        return

    is_selected[cnt] = True
    divide(cnt + 1)
    is_selected[cnt] = False
    divide(cnt + 1)


divide(1)
if result == sys.maxsize:
    print(-1)
else:
    print(result)

File: algorithm/typescript/BOJ17471.ts
// problem : https://www.acmicpc.net/problem/17471

import fs = require("fs");

const input: string[] = fs.readFileSync("/dev/stdin").toString().split("\n");
const N: number = Number(input[0]);
const peopleCount: number[] = input[1].split(" ").map(Number);
const isSelected: boolean[] = new Array(N + 1).fill(false);
const adjMatrix: number[][] = new Array();
let answer: number = Number.MAX_SAFE_INTEGER;

for (let i = 0; i <= N; i++) {
  const row: number[] = new Array(N + 1).fill(0);
  adjMatrix.push(row);
}

for (let i = 2; i <= N + 1; i++) {
  const data: number[] = input[i].split(" ").map(Number);
  for (let j = 1; j <= data[0]; j++) {
    const col: number = data[j];
    adjMatrix[i - 1][col] = adjMatrix[col][i - 1] = 1; // 인접행렬 초기화
  }
}

function BFS(start: number, visited: boolean[]): void {
  let queue: number[] = new Array();
  queue.push(start);
  visited[start] = true;

  while (queue.length != 0) {
    let index: number = queue.shift();

    for (let i = 1; i <= N; i++) {
      if (!visited[i] && adjMatrix[index][i] === 1 && isSelected[start] === isSelected[i]) {
        queue.push(i);
        visited[i] = true;
      }
    }
  }
}

function divide(cnt: number, total: number): void {
  if (cnt === N + 1) {
    if (total === N || total === 0) {
      return;
    }
    let blueStart: number = 0;
    let redStart: number = 0;

    for (let i = 1; i <= N; i++) {
      if (isSelected[i]) {
        blueStart = i;
        break;
      }
    }

    for (let i = 1; i <= N; i++) {
      if (!isSelected[i]) {
        redStart = i;
        break;
      }
    }

    const visited: boolean[] = new Array(N + 1).fill(false);

    BFS(blueStart, visited);
    BFS(redStart, visited);

    for (let i = 1; i <= N; i++) {
      if (!visited[i]) {
        return;
      }
    }

    let redPeople: number = 0;
    let bluePeople: number = 0;

    for (let i = 1; i <= N; i++) {
      if (isSelected[i]) {
        bluePeople += peopleCount[i - 1];
      } else {
        redPeople += peopleCount[i - 1];
      }
    }

    answer = Math.min(answer, Math.abs(bluePeople - redPeople));

    return;
  }

  isSelected[cnt] = true;
  divide(cnt + 1, total + 1);
  isSelected[cnt] = false;
  divide(cnt + 1, total);
}

divide(1, 0);
console.log(answer === Number.MAX_SAFE_INTEGER ? -1 : answer);
